pennylane/qaoa/mixers.py

```diff
@@ -112,3 +114,85 @@ def xy_mixer(graph):
         obs.append(qml.PauliY(node1) @ qml.PauliY(node2))
 
     return qml.Hamiltonian(coeffs, obs)
+
+
+def bit_flip_mixer(graph, n):
+    r"""Creates a bit-flip mixer Hamiltonian.
+
+    This mixer is defined as:
+
+    .. math:: H_M \ = \ \displaystyle\sum_{v \in V(G)} \frac{1}{2^d(v)} X_{v}
+              \displaystyle\prod_{w \in \text{nbhd}(v)} (\mathbb{I} \ + \ (-1)^n Z_w)
+
+    where :math:`V(G)` is the set of vertices of some graph :math:`G`, :math:`d(v)` is the
+    `degree <https://en.wikipedia.org/wiki/Degree_(graph_theory)>`__ of vertex :math:`v`, and
+    :math:`\text{nbhd}(v)` is the `neighbourhood <https://en.wikipedia.org/wiki/Neighbourhood_(graph_theory)>`__
+    of vertex :math:`v`. In addition, :math:`\mathbb{I}` is the identity operator, and :math:`X_v` and :math:`Z_v`
+    are the Pauli-Z and Pauli-X operators on vertex :math:`v`, respectively.
+
+    This mixer was introduced in *From the Quantum Approximate Optimization Algorithm
+    to a Quantum Alternating Operator Ansatz* by Stuart Hadfield, Zhihui Wang, Bryan O'Gorman,
+    Eleanor G. Rieffel, Davide Venturelli, and Rupak Biswas [`arXiv:1709.03489 <https://arxiv.org/abs/1709.03489>`__].
+
+    Args:
+         graph (nx.Graph): A graph defining the collections of wires on which the Hamiltonian acts
+         n (int): Either :math:`0` or :math:`1`. Determines whether the mixer performs a conditional
+                  bit-flip of a vertex if all neighbouring vertices are labelled :math:`0` or :math:`1`.
```

Comment:
I found this one less clear to understand, to me it looks like the choice of n determines whether we plus or minus Z_w. But this is not a bit flip?

Reply:
In the case that n = 0, we have a product over (1 + Z) acting on neighbouring nodes. This will be 0 if there is some neighbouring wire that is labelled with 1 (since (1 + Z|1>) = (1 - 1) = 0), so the state of vertex v will not be flipped by the X_v operation unless all neighbouring nodes are in the |0> state. In the other case, we have (1 - Z) acting on all neighbouring nodes, so all neighbouring nodes have to be in the |1> state in order for v to be flipped.